src/AWS/Shared/Storage/DynamoDBStorage.cs

```diff
@@ -163,34 +165,143 @@ private async Task CreateTable(string tableName, List<KeySchemaElement> keys, Li
+                    tableDescription = await TableWaitOnStatusAsync(tableDescription.TableName, TableStatus.UPDATING, TableStatus.ACTIVE);
+                }
 
-                    await Task.Delay(2000);
+                await TableUpdateTtlAsync(tableDescription.TableName, ttlAttributeName);
 
-                } while (description.TableStatus == TableStatus.CREATING);
+                var existingGlobalSecondaryIndexes = tableDescription.GlobalSecondaryIndexes.Select(globalSecondaryIndex => globalSecondaryIndex.IndexName).ToArray();
+                var secondaryIndexesToCreate = (secondaryIndexes ?? Enumerable.Empty<GlobalSecondaryIndex>()).Where(secondaryIndex => !existingGlobalSecondaryIndexes.Contains(secondaryIndex.IndexName));
 
-                if (!string.IsNullOrEmpty(ttlAttributeName))
+                foreach (var secondaryIndex in secondaryIndexesToCreate)
                 {
-                    await ddbClient.UpdateTimeToLiveAsync(new UpdateTimeToLiveRequest
-                    {
-                        TableName = tableName,
-                        TimeToLiveSpecification = new TimeToLiveSpecification { AttributeName = ttlAttributeName, Enabled = true }
-                    });
+                    await TableCreateSecondaryIndex(tableDescription.TableName, attributes, secondaryIndex);
                 }
-                if (description.TableStatus != TableStatus.ACTIVE)
-                    throw new InvalidOperationException($"Failure creating table {tableName}");
             }
             catch (Exception exc)
             {
-                Logger.Error(ErrorCode.StorageProviderBase, $"Could not create table {tableName}", exc);
+                Logger.Error(ErrorCode.StorageProviderBase, $"Could not update table {tableDescription.TableName}", exc);
                 throw;
             }
         }
 
+        private async Task TableCreateSecondaryIndex(string tableName, List<AttributeDefinition> attributes, GlobalSecondaryIndex secondaryIndex)
+        {
+            await ddbClient.UpdateTableAsync(new UpdateTableRequest
+            {
+                TableName = tableName,
+                GlobalSecondaryIndexUpdates = new List<GlobalSecondaryIndexUpdate>
+                {
+                    new GlobalSecondaryIndexUpdate
+                    {
+                        Create = new CreateGlobalSecondaryIndexAction()
+                        {
+                            IndexName = secondaryIndex.IndexName,
+                            Projection = secondaryIndex.Projection,
+                            ProvisionedThroughput = secondaryIndex.ProvisionedThroughput,
+                            KeySchema = secondaryIndex.KeySchema
+                        }
+                    }
+                },
+                AttributeDefinitions = attributes
+            });
+
+            await TableIndexWaitOnStatusAsync(tableName, secondaryIndex.IndexName, IndexStatus.CREATING, IndexStatus.ACTIVE);
```

Comment:
I'm not certain if it's necessary, but in some similar code, we also check that the table itself has gone from updating to active.

Reply:
It could complicate things, but at the cost of an extra description request, it could also just call `TableWaitOnStatusAsync`.
I unfortunately don't recall if the table status actually changed when adding a GSI to know if it's necessary.

Combining vs leaving separate seems like a wash, but maybe I'm missing something.

Reply:
> It could complicate things, but at the cost of an extra description request, it could also just call `TableWaitOnStatusAsync`. I unfortunately don't recall if the table status actually changed when adding a GSI to know if it's necessary.
> 
> Combining vs leaving separate seems like a wash, but maybe I'm missing something.

Doing the above would basically guarantee that the tables has gone  through an update. I'll check if adding a GSI triggers a table status change.